internals/overlord/logstate/syslog/syslog.go

```diff
@@ -0,0 +1,271 @@
+	if len(labels) == 0 {
+		delete(c.labels, serviceName)
+		return
+	}
+	var buf bytes.Buffer
+
+	if len(c.options.SDID) > 0 {
+		fmt.Fprintf(&buf, "[%s@%d", c.options.SDID, canonicalPrivEnterpriseNum)
+	} else {
+		fmt.Fprintf(&buf, "[%s@%d", "pebble", canonicalPrivEnterpriseNum)
+	}
+
+	// Sort label keys to get deterministic output
+	keys := make([]string, 0, len(labels))
+	for key := range labels {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		fmt.Fprintf(&buf, " %s=\"", key)
+		value := labels[key]
+		// escape the value according to RFC5424 6.3.3
+		for i := 0; i < len(value); i++ {
+			// don't use "for _, c := range value" as we don't want runes
+			c := value[i]
+			if c == '"' || c == '\\' || c == ']' {
+				buf.WriteByte('\\')
+			}
+			buf.WriteByte(c)
+		}
+		buf.WriteByte('"')
+	}
+	buf.WriteByte(']')
+	c.labels[serviceName] = buf.String()
+}
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if len(c.entries) >= c.options.MaxRequestEntries {
+		// 'entries' is full - remove the first element to make room
+		// Zero the removed element to allow garbage collection
+		c.entries[0] = entryWithService{}
+		c.entries = c.entries[1:]
+	}
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer
+		c.entries = c.buffer[:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection
+		for i := len(c.entries); i < len(c.buffer); i++ {
+			c.buffer[i] = entryWithService{}
+		}
+	}
+
+	entry.Message = strings.TrimSuffix(entry.Message, "\n")
+
+	c.entries = append(c.entries, entryWithService{
+		Timestamp: entry.Time.Format(time.RFC3339Nano), // Format: 2021-05-26T12:37:01Z
+		Message:   entry.Message,
+		service:   entry.Service,
+	})
+	return nil
+}
+
+func (c *Client) ensureConnected(ctx context.Context) error {
+	if c.conn != nil {
+		return nil
+	} else if c.closed {
+		return fmt.Errorf("write to closed SyslogBackend")
+	}
+
+	d := net.Dialer{Timeout: c.options.DialTimeout}
+	conn, err := d.DialContext(ctx, c.location.Scheme, c.location.Host)
+	if err != nil {
+		return fmt.Errorf("cannot connect to %s", c.location)
+	}
+
+	c.conn = conn
+	return nil
+}
+
+func (c *Client) Close() error {
+	c.closed = true
+	if c.conn == nil {
+		return nil
+	}
+
+	err := c.conn.Close()
+	c.conn = nil
+	return err
+}
+
+// Flush sends buffered logs to the syslog endpoint.
+func (c *Client) Flush(ctx context.Context) error {
+	if len(c.entries) == 0 {
+		return nil
+	}
+
+	err := c.ensureConnected(ctx)
+	if err != nil {
+		return err
+	}
+	c.sendBuf.Reset()
+
+	for _, entry := range c.entries {
+		structuredData, ok := c.labels[entry.service]
+		if !ok {
+			structuredData = "-"
+		}
+
+		defaultPriority := priorityVal(facilityUserLevelMessage, severityNotice)
+
+		// Build the message first to calculate length
+		var msgBuf bytes.Buffer
```

Comment:
This wasn't what I had in mind at all, as it doesn't actually reduce allocations. It avoids fmt.Sprintf, but we're still doing an extra allocation by building this new bytes.Buffer each time around the loop -- that's what the `sendBuf` is for.

See let's use something closer to my [original suggestion in this comment](https://github.com/canonical/pebble/pull/725#discussion_r2548126996). We can also get rid of the `strconv.Itoa`, as those (normally) cause a memory allocation too.

[See a benchmark using `go test -bench` here.](https://github.com/jy5275/pebble/pull/1)

```
BenchmarkEncodeEntryOld-16    	 8141732	       124.0 ns/op	     195 B/op	       3 allocs/op
BenchmarkEncodeEntry-16       	49393112	        24.05 ns/op	       0 B/op	       0 allocs/op
```